state/execution.go

```diff
@@ -296,51 +299,54 @@ func execBlockOnProxyApp(
 	return abciResponses, nil
 }
 
-func getBeginBlockValidatorInfo(block *types.Block, lastValSet *types.ValidatorSet, stateDB dbm.DB) (abci.LastCommitInfo, []abci.Evidence) {
-
-	// Sanity check that commit length matches validator set size -
-	// only applies after first block
+func getBeginBlockValidatorInfo(block *types.Block, stateDB dbm.DB) (abci.LastCommitInfo, []abci.Evidence) {
+	voteInfos := make([]abci.VoteInfo, len(block.LastCommit.Precommits))
+	byzVals := make([]abci.Evidence, len(block.Evidence.Evidence))
 	if block.Height > 1 {
+		lastValSet, err := LoadValidators(stateDB, block.Height-1)
+		if err != nil {
+			panic(err) // shouldn't happen
+		}
+
+		// Sanity check that commit length matches validator set size -
+		// only applies after first block
+
 		precommitLen := len(block.LastCommit.Precommits)
 		valSetLen := len(lastValSet.Validators)
 		if precommitLen != valSetLen {
 			// sanity check
 			panic(fmt.Sprintf("precommit length (%d) doesn't match valset length (%d) at height %d\n\n%v\n\n%v",
 				precommitLen, valSetLen, block.Height, block.LastCommit.Precommits, lastValSet.Validators))
 		}
-	}
 
-	// Collect the vote info (list of validators and whether or not they signed).
-	voteInfos := make([]abci.VoteInfo, len(lastValSet.Validators))
-	for i, val := range lastValSet.Validators {
-		var vote *types.CommitSig
-		if i < len(block.LastCommit.Precommits) {
-			vote = block.LastCommit.Precommits[i]
+		for i, val := range lastValSet.Validators {
+			var vote *types.CommitSig
+			if i < len(block.LastCommit.Precommits) {
+				vote = block.LastCommit.Precommits[i]
+			}
+			voteInfo := abci.VoteInfo{
+				Validator:       types.TM2PB.Validator(val),
+				SignedLastBlock: vote != nil,
+			}
+			voteInfos[i] = voteInfo
 		}
-		voteInfo := abci.VoteInfo{
-			Validator:       types.TM2PB.Validator(val),
-			SignedLastBlock: vote != nil,
+
+		for i, ev := range block.Evidence.Evidence {
```

Comment:
This should be outside the `if block.Height > 1` as well - can't we include evidence in the first block?